src/gdcmodels/mapper.py

```diff
@@ -0,0 +1,119 @@
+
+    Args:
+        mapping: The elasticsearch mapping or property to walk.
+        path: The path which has been walked upto the current mapping or property.
+
+    Yields:
+        The path and its associated property definition.
+    """
+    for name, property in mapping.get("properties", {}).items():
+        sub_path = f"{path}.{name}"
+        yield sub_path, property
+        yield from _walk_mapping(property, sub_path)
+
+
+class ModelMapper:
+    """This translates the model into elasticsearch mappings/settings.
+
+    NOTE: Instances of this class should be instantiated via `get_es_models`
+    """
+
+    __slots__ = ("_index_name", "_doc_type", "_settings", "_mapping")
+
+    def __init__(
+        self,
+        index_name: str,
+        doc_type: str,
+        settings: Mapping[str, Any],
+        mapping: esmodels.ESMapping,
+    ) -> None:
+        self._index_name = index_name
+        self._doc_type = doc_type
+        self._settings = settings
+        self._mapping = mapping
+
+    @property
+    def index_name(self) -> str:
+        return self._index_name
+
+    @property
+    def doc_type(self) -> str:
+        return self._doc_type
+
+    @property
+    def mappings(self) -> esmodels.ESMapping:
+        """The elasticsearch mappings associated with the model."""
+        return self._mapping
+
+    @property
+    def settings(self) -> Mapping[str, Any]:
+        """The settings object for the elasticsearch index."""
+        return self._settings
+
+    def select_mapping(
+        self,
+        doc_type: str,
+        selector: Optional[Selector] = None,
+    ) -> Union[esmodels.ESMapping, esmodels.Property]:
+        """Select a sub-mapping or the entire mapping based on the doc_type.
+
+        Using the given doc_type any property within this mapping with that name will
+        have it's property definition returned. If multiple exist a selector should be
+        insure that the selection is deterministic.
+
+        Example:
+            # Just return the very first path:
+            mapper.select_mapping(
+                "annotation",
+                selector=functools.partial(more_itertools.take, 1)
+            )
+            # Only select root.(...).cnv.(...).observation:
+            mapper.select_mapping("observation", selector="cnv")
+
+        Args:
+            doc_type: name of the property (or the index itself) which should be
+                selected.
+            selector: The selector can either be a callable which takes in all possible
+                paths to the given doc type which it filters to the desired path or the
+                name of a parent property that must be within the desired properties
+                path.
+
+        Returns:
+            The entire mapping definition or the property details associated with the
+            given doc_type.
+        """
+        if self.doc_type in [doc_type, f"{doc_type}_centric"]:
```

Comment:
Trying to understand why we have the extra `f"{doc_type}_centric"` item in the list. If `self.doc_type = case_centric`, for a particular instance passing a `doc_type = case` to this method returns the mapping for the `case_centric`. Is this the intended behavior?

Reply:
For that you'd have to look to [here](https://github.com/NCI-GDC/gdc-mutation-indexer/blob/026db7739b0dd59dafdfbc40384880abe9e35e7e/src/mutation_indexer/builders/cnv_centric.py#L58) and [here](https://github.com/NCI-GDC/gdc-mutation-indexer/blob/026db7739b0dd59dafdfbc40384880abe9e35e7e/src/mutation_indexer/builders/ssm_centric.py#L60-L62). Unfortunately, the entire [model mapper](https://github.com/NCI-GDC/normalizer/blob/382d9cf328d098f0c889c04d9d249251f58c7942/normalizer/mapper.py#L146) is actually a structure that has its roots in MI. This logic is ultimately used to dynamically build a bunch of stuff in the [df_builders](https://github.com/NCI-GDC/gdc-mutation-indexer/blob/develop/src/mutation_indexer/builders/df_builders.py) via the [utils](https://github.com/NCI-GDC/gdc-mutation-indexer/blob/026db7739b0dd59dafdfbc40384880abe9e35e7e/src/mutation_indexer/builders/utils.py#L100-L111) module which, as you most certainly pointed out in your other comment, can be hard to read/understand/know what is coming out of those functions and know that they are working as intended. To add insult to injury, those builders are also SUPER inefficient compared to the actual builders in MI as they each rebuild basically the same data multiple times. 

These things all said for the time being we have to maintain this legacy interface until I (or someone else) gets the cycles to go switch the logic in MI to behave in a more predictable/stable/efficient way.